### DISCUSSION TITLE
Where and how to use `{{securecontext_inline}}`

### DISCUSSION THREAD

**Discussion by skyclouds2001** [mdn/mdn-community#482] [Content] [answered]
Currently, there are two ways to indicate that a page is only available in Secure Context ———— `{{securecontext_inline}}` and `{{securecontext_header}}`. They are used for same purpose but for different situations in documentation.

But the fact is, the usage between `{{securecontext_header}}` and `{{securecontext_inline}}` is every different: `{{securecontext_header}}` widely used but `{{securecontext_inline}}` is rawly used. Also, `{{securecontext_header}}` is mentioned in Writing guide pages but `{{securecontext_inline}}` not. Besides, `{{securecontext_inline}}` its size is quite big and a little not pretty, compared with other inline-type macros ———— `{{non-standard_inline}}` `{{deprecated_inline}}` `{{experimental_inline}}` `{{readonlyinline}}` `{{optional_inline}}`. This makes it is more likely to break the line then other inline-type macros. And using this maight be confused with `{{securecontext_header}}` which could express the same meaning. So, personnally now I'd prefer not to use `{{securecontext_inline}}` in documentation writing.

I wonder if it is better to also use `{{securecontext_inline}}` to indicate secure context with `{{securecontext_header}}`, and it is necessary to express the secure context requirement to readers in situations that might be fit to use `{{securecontext_inline}}` (like API's interfaces in *API summary*, interface's property in *interface summary*).

> some examples that use `{{securecontext_inline}}`
> * https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API
> * https://developer.mozilla.org/en-US/docs/Web/API/Geolocation
> * https://developer.mozilla.org/en-US/docs/Web/API/GeolocationCoordinates
> * https://developer.mozilla.org/en-US/docs/Web/API/GeolocationPosition
> * https://developer.mozilla.org/en-US/docs/Web/API/GeolocationPositionError

**Comment by hamishwillee** [comment]
I don't think of we need `{{securecontext_inline}}`. My reasoning is that unlike deprecated/experimental/non-standard you're not highlighting a decision as to whether something _should_ be used, but how it is used. You're still going to have to use it if you need the feature, and you can find that about the secure context from the top level API or the interface level. You shouldn't usually need to know about individual interfaces needing the context from that top level. 

But if we do decide it is useful there should be a guide on this, and it should be applied consistently everywhere.

**Comment by teoli2003** [comment]
I don't know if there are such cases, but I would say the only cases where it would be useful is if an interface is NOT restricted to secure contexts, but one of its members (a method or a property) is.

But I don't have such an example (except on `Window` but these really are global, and we may want to exclude them).

Note: even in such cases, using the header macro on the member page may be enough.

**Reply by hamishwillee** [reply]
> I don't know if there are such cases, but I would say the only cases where it would be useful is if an interface is NOT restricted to secure contexts, but one of its members (a method or a property) is.

Good points. I think Window might be the only place I would find that useful.

**Reply by teoli2003** [reply]
Of course, we should fix the look of the inline macro to be very similar to the other inline macros.

We should eliminate the macro if we decide not to add inline secure context macros.

**Reply by skyclouds2001** [reply]
like `Screen` interface that extended via [Window Management API](https://www.w3.org/TR/window-management/) ?

**Reply by teoli2003** [reply]
Yes, these two members should have a `securecontext_inline` macro. (But we should uniformize their look with the other inline macros).

So, I think that we should :
- update the [macro](https://github.com/mdn/yari/blob/main/kumascript/macros/secureContext_inline.ejs) on mdn/yari to have the same look and fill (that is, using `badge` instead of `notecard`)
- update the meta-documentation
- add `{{securecontext_inline}}` to the few needed cases.

Let's wait for the other maintainers' opinion.

**Reply by wbamberg** [reply]
Is the suggestion here that we only have the badge when some but not all of an interface's members are secure context required? For instance, we would not badge the members of https://developer.mozilla.org/en-US/docs/Web/API/GeolocationPosition, but we would badge the (appropriate) members of `Window`?

Would that be confusingly inconsistent?

**Reply by teoli2003** [reply]
Yes, this is my proposal.

It would be inconsistent, but it is what specs are doing: they set the SecureContext flag on the interface, or on members.
`Screen` is not restricted to secure context, but some of its members are:
![Capture d’écran 2023-11-27 à 20 52 45](https://github.com/mdn/content/assets/1466293/03bcfdfa-a1a3-47b9-9448-f06ab5a14966)
The whole `GeolocationPosition` is restricted, so the spec only marks it on the interface:
![Capture d’écran 2023-11-27 à 20 58 25](https://github.com/mdn/content/assets/1466293/913d0fb5-d735-4f18-ab6e-98d4831caee8)


I don't think it will be confusing for users, though:
- either devs will look for the interface, and there will be a banner at the top of the interface if it is secure-context only. (There will also be a banner at the top of the member page if they follow the link.)
- or devs will go directly to the member page, and there will be a banner at the top of the page.

However, I've no strong objection to having the inline banner everywhere, like this (though the inline badge should be fixed, of course)
![Capture d’écran 2023-11-27 à 21 04 35](https://github.com/mdn/content/assets/1466293/1cd3a1b7-e7b6-4123-b9e3-1f77f072a20e)

**Reply by wbamberg** [reply]
What's worse though is that it's not only self-inconsistent, it's inconsistent with the other ways we use headers and badges, like non-standard. Someone might think, why do we always indicate non-standard with a badge, but only sometimes indicate secure context? But maybe it's OK. I don't like the alternative, of always badging members, very much either!

**Reply by hamishwillee** [reply]
No one is ever going to be happy here. Note, one day we might want to auto do this through IDL metadata, so we should consider what that might look like. 

Personally I'm OK with 
- tagging interfaces with the header if everything is secure context
- individually tagging properties/methods if only parts are secure context/
- Using a badge! 

As it is now I'd prefer we also had the header on property/method pages too, but I'd rather that was some kind of badge too, generated from metadata, and not a big fat secure context note.

I am OK if we are not consistent with deprecated etc, as long as we are consistent in the category. secure context is not the same as deprecated etc.

**Reply by teoli2003** [reply]
> What's worse though, is that it's not only self-inconsistent, it's inconsistent with the other ways we use headers and badges, like non-standard.

I think this is a good point! That way, we have one rule for all these badges/icons. 

So what about:
- using a badge (I think there is consensus on this)
- tagging interfaces with the header macro if everything is secure context (I think there is consensus on this)
- individually tagging members if they are secure context only.

> As it is now, I'd prefer we also had the header on property/method pages too, but I'd rather that was some kind of badge too, generated from metadata and not a big fat secure context note.

A potential mitigating issue here would be to put the banner not at the top of the page, but below the description. But this is probably a separate discussion: we should think about how to do all our banners differently (especially when there are several banners: we may want to group some). As this would likely touch the _baseline_ feature, I would like to delay this discussion: there are too many moving parts for the moment.

> Note, one day we might want to auto do this through IDL metadata, so we should consider what that might look like.

Absolutely! Experimental/deprecated/non-standard icons and banners are scripted for coherence with BCD; read-only and secure-context badges (and banners) can be scripted to be coherent with WebIDL (via w3c/webref).

Having a strict rule (whichever it is), is the necessary first step toward it. 😈

**Reply by hamishwillee** [reply]
> using a badge (I think there is consensus on this)

We have consensus that we want to use a badge with consistent appearance.

> tagging interfaces with the header macro if everything is secure context (I think there is consensus on this)
> individually tagging members if they are secure context only.

I'm not certain this is unambiguous. I think you are saying:
- In interface page
  - if whole interface is secure context
    - tag the page with header macro
    - don't tag the individual members with badge
  - if only part of the interface is secure context
    - don't tag the page with header macro
    - tag the individual members with badge
- In property/method page
  - tag the page with header macro if it is secure context and not otherwise. Irrespective of the interface.

If not, can you expand?

**Reply by teoli2003** [reply]
Yes, I think the only point of disagreement is:

In an interface page

**Alternative 1**:
 - if the whole interface is secure context
   - tag the page with a header macro
      - don't tag the individual members with badge
      
  vs
  
 **Alternative 2**:
  - if the whole interface is secure context
   - tag the page with a header macro
      - also tag the individual members with badge

I have no strong feelings here.

Alternative 1:
Pro: no Christmas tree
Con: when you are on a long page, you don't know which kind of interface page you are.

Alternative 2:
Pro:
- consistency between all interface pages (without looking at the top of the page)
- consistency with other badges on every interface page

Con: verbose (a bit of a Christmas tree)



I would prefer alternative 1, but I am neutral about alternative 2 (I will not oppose it).

**Reply by skyclouds2001** [reply]
I also prefer with *Alternative 1*

> * if the whole interface is secure context
>   
>   * tag the page with a header macro
>     
>     * don't tag the individual members with badge

because if an interface requires secure context, then its methods or properties definitely requires secure context, but if an interface does not require secure context, then its methods or properties may still require secure context

**Reply by skyclouds2001** [reply]
> > using a badge (I think there is consensus on this)
> 
> We have consensus that we want to use a badge with consistent appearance.
> 
> > tagging interfaces with the header macro if everything is secure context (I think there is consensus on this)
> > individually tagging members if they are secure context only.
> 
> I'm not certain this is unambiguous. I think you are saying:
> 
> * In interface page
>   
>   * if whole interface is secure context
>     
>     * tag the page with header macro
>     * don't tag the individual members with badge
>   * if only part of the interface is secure context
>     
>     * don't tag the page with header macro
>     * tag the individual members with badge
> * In property/method page
>   
>   * tag the page with header macro if it is secure context and not otherwise. Irrespective of the interface.
> 
> If not, can you expand?

Another possible situation is for api landing page (also constructor/event pages, but they could combine with property/method pages)

**Reply by hamishwillee** [reply]
I prefer Alternative 1 too. Sounds like broad consensus there.

What about property/method pages? They always get the secure context header (if require secure context)?

**Reply by teoli2003** [reply]
Yes, they should: there is no guarantee that a reader will read the interface page first.

**Comment by wbamberg** [comment]
It looks to me like you've mixed up the proposals and the pros/cons?

> **Alternative 1**:
> ...
>         * don't tag the individual members with badge
> ...
> Alternative 1: 
> ...
> 
> Con: verbose (a bit of a Christmas tree)

**Reply by skyclouds2001** [reply]
yes, my mistake

**Reply by teoli2003** [reply]
I mixed and corrected.

**Comment by hamishwillee** [comment, marked as answer]
Consensus?
- 

1. Fix up the secure-inline macro to be a badge.
2. In an interface page (Alternative 1)

   - if the whole interface is secure context
     - tag the page with a header macro
     - don't tag the individual members with badge
   - if only part of the interface is secure context
     - don't tag the page with header macro
     - tag the individual members with badge
 
3. Tag property/method pages with secure context, irrespective of the interface.
4. The API overview should have a header too if the whole thing is secure context. Not sure about what it should do if "partially secure context". I guess same rule as for interfaces.

**Reply by OnkarRuikar** [reply]
:+1: alternative 1

Do we know any case where only part of an interface is secure?
If no then we don't need the inline secure context badge.

Has BCD been maintaining secure context flag for each feature?

**Reply by teoli2003** [reply]
Yes, the [`BaseAudioContext`](https://developer.mozilla.org/en-US/docs/Web/API/BaseAudioContext) interface is an example where only a member is restricted to secure context.

No, BCD will not maintain it, but `w3c/webref` has this information. It is an extended attribute in the Webidl, `[SecureContext]`, that can be set at the interface level or on a member:
```
interface BaseAudioContext : EventTarget {
    readonly attribute AudioDestinationNode destination];
    ...
    [SameObject, SecureContext] readonly attribute AudioWorklet audioWorklet];
    attribute EventHandler onstatechange;
    ...
}
```

**Reply by Rumyra** [reply]
I'm ok with this proposal 👍 

(Apologies I didn't bring it up yesterday @teoli2003 )

**Reply by hamishwillee** [reply]
So, who can fix the badge? Do we have enough "approval" to update our docs with this as a guideline? I'd say yes.

**Reply by teoli2003** [reply]
Yes, it looks like we have only approvals and nobody against it, which is the definition of a consensus, so we can move forward.

**Reply by teoli2003** [reply]
I created the yari PR to fix the look of the badge: https://github.com/mdn/yari/pull/10214

**Reply by skyclouds2001** [reply]
Current usage of this macros:

except those will remove via #31302 

- `BaseAudioContext.audioWorklet`
- `Crypto.subtle`
- `CSS.paintWorklet`
- `Screen.isExtended` `Screen/change_event`
- `Window.getScreenDetails()`

- Clipboard API
- Sensor APIs
- Window Management API

**Comment by OnkarRuikar** [comment]
> So, who can fix the badge? Do we have enough "approval" to update our docs with this as a guideline? I'd say yes.

There is a "Mozilla-wide Production Change Freeze" [till Jan 2](https://discord.com/channels/1009925603572600863/1170042997212184576/1184812856072486933) so till then the inline badge can't be fixed. 
Any idea how the docs can be updated? May be we can write a script to fetch the statues from `webref` repo. Or do it manually if there are only few features that require securecontext.

**Reply by hamishwillee** [reply]
I think the first step would be to update the macro and template docs with "the rule" so that we all know what we're doing going forward.

W.r.t. "how", I think a script would make sense. I bet @teoli2003 has some ideas about how to get started :-)

**Reply by teoli2003** [reply]
I agree we should update the macro and the meta docs first.

If the macro change is only merged on Jan 2, that's not a problem: we use the inline macro only in specific cases that often already use the ugly look of the macro.

Then we need to update, ideally with a script that can run weekly afterward.

**Comment by OnkarRuikar** [comment]
Does the decision apply to all the inline macros, e.g. status macros `{{Experimental_inline}}`?

**Reply by OnkarRuikar** [reply]
It was decided to have [a separate discussion](https://github.com/orgs/mdn/discussions/654) for inline status macros.
